File: src/api/training.ts
export type TrainingMode = "light" | "standard" | "fine";
export type TrainingModel = "flux-schnell" | "flux-dev" | "sdxl";
export type TrainingBusinessStatus =
  | "draft"
  | "queued"
  | "preparing"
  | "training"
  | "sampling"
  | "validating"
  | "completed"
  | "failed"
  | "canceled";
export type TrainingQueueStatus = "pending" | "running" | "completed" | "failed";

export type TrainingImage = {
  progress?: number;
  view?: string;
  image: string;
};

export type TrainingTaskSnapshot = {
  id: string;
  characterId: string;
  queueTaskId: string;
  requestedMode: TrainingMode;
  effectiveMode: TrainingMode;
  requestedModel: TrainingModel;
  effectiveModel: TrainingModel;
  strategyDefaultModel: string;
  runtimeDefaultModel: string;
  requestedSteps: number;
  effectiveSteps: number;
  requestedRank: number;
  effectiveRank: number;
  precheck: {
    vramGB: number;
    source: string;
    result: string;
  };
  downgradeReasons: Array<{ code: string; message: string }>;
  config: Record<string, unknown>;
  businessStatus: TrainingBusinessStatus;
  queueStatus: TrainingQueueStatus;
  progress: number;
  currentStage: string;
  latestMessage: string | null;
  latestError: string | null;
  userVisibleError: string | null;
  samplePreviews: TrainingImage[];
  validationImages: TrainingImage[];
  retrainOfTaskId: string | null;
  createdAt: string;
  updatedAt: string;
  startedAt: string | null;
  finishedAt: string | null;
};

export type StartTrainingPayload = {
  characterId: string;
  mode: TrainingMode;
  baseModel?: TrainingModel;
  confirmFluxDevLicense?: boolean;
  retrainOfTaskId?: string;
  retrainStepDelta?: number;
};

const DEFAULT_API_BASE_URL = "http://127.0.0.1:8000";

function resolveTrainingTasksUrl(characterId?: string): string {
  if (characterId) {
    return `${DEFAULT_API_BASE_URL}/api/training/tasks?characterId=${encodeURIComponent(characterId)}`;
  }
  return `${DEFAULT_API_BASE_URL}/api/training/tasks`;
}

function resolveTrainingTaskUrl(taskId: string): string {
  return `${DEFAULT_API_BASE_URL}/api/training/tasks/${encodeURIComponent(taskId)}`;
}

function resolveTrainingStartUrl(): string {
  return `${DEFAULT_API_BASE_URL}/api/training/start`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isTrainingMode(value: unknown): value is TrainingMode {
  return value === "light" || value === "standard" || value === "fine";
}

function isTrainingModel(value: unknown): value is TrainingModel {
  return value === "flux-schnell" || value === "flux-dev" || value === "sdxl";
}

function isTrainingBusinessStatus(value: unknown): value is TrainingBusinessStatus {
  return (
    value === "draft" ||
    value === "queued" ||
    value === "preparing" ||
    value === "training" ||
    value === "sampling" ||
    value === "validating" ||
    value === "completed" ||
    value === "failed" ||
    value === "canceled"
  );
}

function isTrainingQueueStatus(value: unknown): value is TrainingQueueStatus {
  return value === "pending" || value === "running" || value === "completed" || value === "failed";
}

function isTrainingImage(value: unknown): value is TrainingImage {
  if (!isRecord(value) || typeof value.image !== "string") {
    return false;
  }
  const hasValidProgress = value.progress === undefined || typeof value.progress === "number";
  const hasValidView = value.view === undefined || typeof value.view === "string";
  return hasValidProgress && hasValidView;
}

function isTrainingTaskSnapshot(value: unknown): value is TrainingTaskSnapshot {
  if (!isRecord(value)) {
    return false;
  }

  if (
    typeof value.id !== "string" ||
    typeof value.characterId !== "string" ||
    typeof value.queueTaskId !== "string" ||
    !isTrainingMode(value.requestedMode) ||
    !isTrainingMode(value.effectiveMode) ||
    !isTrainingModel(value.requestedModel) ||
    !isTrainingModel(value.effectiveModel) ||
    typeof value.strategyDefaultModel !== "string" ||
    typeof value.runtimeDefaultModel !== "string" ||
    typeof value.requestedSteps !== "number" ||
    typeof value.effectiveSteps !== "number" ||
    typeof value.requestedRank !== "number" ||
    typeof value.effectiveRank !== "number" ||
    !isRecord(value.precheck) ||
    typeof value.precheck.vramGB !== "number" ||
    typeof value.precheck.source !== "string" ||
    typeof value.precheck.result !== "string" ||
    !Array.isArray(value.downgradeReasons) ||
    !isRecord(value.config) ||
    !isTrainingBusinessStatus(value.businessStatus) ||
    !isTrainingQueueStatus(value.queueStatus) ||
    typeof value.progress !== "number" ||
    typeof value.currentStage !== "string" ||
    !(value.latestMessage === null || typeof value.latestMessage === "string") ||
    !(value.latestError === null || typeof value.latestError === "string") ||
    !(value.userVisibleError === null || typeof value.userVisibleError === "string") ||
    !Array.isArray(value.samplePreviews) ||
    !Array.isArray(value.validationImages) ||
    !(value.retrainOfTaskId === null || typeof value.retrainOfTaskId === "string") ||
    typeof value.createdAt !== "string" ||
    typeof value.updatedAt !== "string" ||
    !(value.startedAt === null || typeof value.startedAt === "string") ||
    !(value.finishedAt === null || typeof value.finishedAt === "string")
  ) {
    return false;
  }

  const reasonsValid = value.downgradeReasons.every(
    (item) => isRecord(item) && typeof item.code === "string" && typeof item.message === "string",
  );

  if (!reasonsValid) {
    return false;
  }

  const sampleValid = value.samplePreviews.every((item) => isTrainingImage(item));
  const validationValid = value.validationImages.every((item) => isTrainingImage(item));
  return sampleValid && validationValid;
}

function extractErrorMessage(payload: unknown): string {
  if (isRecord(payload) && typeof payload.detail === "string") {
    return payload.detail;
  }
  return "训练请求失败，请稍后重试。";
}

async function parseJsonOrThrow(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch {
    throw new Error("INVALID_TRAINING_RESPONSE");
  }
}

export async function fetchTrainingTasks(
  characterId?: string,
  signal?: AbortSignal,
): Promise<TrainingTaskSnapshot[]> {
  let response: Response;
  try {
    response = await fetch(resolveTrainingTasksUrl(characterId), { signal });
  } catch {
    throw new Error("TRAINING_SERVICE_UNAVAILABLE");
  }

  const payload = await parseJsonOrThrow(response);

  if (!response.ok) {
    throw new Error(extractErrorMessage(payload));
  }

  if (!Array.isArray(payload) || !payload.every((item) => isTrainingTaskSnapshot(item))) {
    throw new Error("INVALID_TRAINING_RESPONSE");
  }

  return payload;
}

export async function fetchTrainingTask(
  taskId: string,
  signal?: AbortSignal,
): Promise<TrainingTaskSnapshot> {
  let response: Response;
  try {
    response = await fetch(resolveTrainingTaskUrl(taskId), { signal });
  } catch {
    throw new Error("TRAINING_SERVICE_UNAVAILABLE");
  }

  const payload = await parseJsonOrThrow(response);

  if (!response.ok) {
    throw new Error(extractErrorMessage(payload));
  }

  if (!isTrainingTaskSnapshot(payload)) {
    throw new Error("INVALID_TRAINING_RESPONSE");
  }

  return payload;
}

export async function startTraining(
  payload: StartTrainingPayload,
  signal?: AbortSignal,
): Promise<TrainingTaskSnapshot> {
  let response: Response;
  try {
    response = await fetch(resolveTrainingStartUrl(), {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal,
    });
  } catch {
    throw new Error("TRAINING_SERVICE_UNAVAILABLE");
  }

  const responsePayload = await parseJsonOrThrow(response);

  if (!response.ok) {
    throw new Error(extractErrorMessage(responsePayload));
  }

  if (!isTrainingTaskSnapshot(responsePayload)) {
    throw new Error("INVALID_TRAINING_RESPONSE");
  }

  return responsePayload;
}

File: backend/app/api/training.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from fastapi import APIRouter, HTTPException, Query, Request, status
from pydantic import BaseModel, ConfigDict, Field

from app.services.training import (
    TrainingCharacterNotFoundError,
    TrainingJobNotFoundError,
    TrainingService,
    TrainingServiceError,
    TrainingStartPayload,
    TrainingValidationError,
)

router = APIRouter()


class TrainingPrecheckPayload(BaseModel):
    vramGB: float = Field(ge=0)
    source: str
    result: str


class TrainingDowngradeReasonPayload(BaseModel):
    code: str
    message: str


class TrainingJobPayload(BaseModel):
    id: str
    characterId: str
    queueTaskId: str
    requestedMode: Literal["light", "standard", "fine"]
    effectiveMode: Literal["light", "standard", "fine"]
    requestedModel: Literal["flux-schnell", "flux-dev", "sdxl"]
    effectiveModel: Literal["flux-schnell", "flux-dev", "sdxl"]
    strategyDefaultModel: str
    runtimeDefaultModel: str
    requestedSteps: int = Field(ge=1)
    effectiveSteps: int = Field(ge=1)
    requestedRank: int = Field(ge=1)
    effectiveRank: int = Field(ge=1)
    precheck: TrainingPrecheckPayload
    downgradeReasons: list[TrainingDowngradeReasonPayload]
    config: dict[str, Any]
    businessStatus: Literal[
        "draft",
        "queued",
        "preparing",
        "training",
        "sampling",
        "validating",
        "completed",
        "failed",
        "canceled",
    ]
    queueStatus: Literal["pending", "running", "completed", "failed"]
    progress: int = Field(ge=0, le=100)
    currentStage: str
    latestMessage: str | None = None
    latestError: str | None = None
    userVisibleError: str | None = None
    samplePreviews: list[dict[str, Any]]
    validationImages: list[dict[str, Any]]
    retrainOfTaskId: str | None = None
    createdAt: str
    updatedAt: str
    startedAt: str | None = None
    finishedAt: str | None = None


class TrainingStartRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    character_id: str = Field(alias="characterId", min_length=1)
    mode: Literal["light", "standard", "fine"]
    base_model: Literal["flux-schnell", "flux-dev", "sdxl"] | None = Field(
        default=None,
        alias="baseModel",
    )
    confirm_flux_dev_license: bool = Field(default=False, alias="confirmFluxDevLicense")
    retrain_of_task_id: str | None = Field(default=None, alias="retrainOfTaskId")
    retrain_step_delta: int | None = Field(default=None, alias="retrainStepDelta")


@dataclass(slots=True)
class TrainingRuntime:
    service: TrainingService


def resolve_training_runtime(request: Request) -> TrainingRuntime:
    service = getattr(request.app.state, "training_service", None)
    if not isinstance(service, TrainingService):
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="训练服务暂不可用，请稍后重试。",
        )
    return TrainingRuntime(service=service)


@router.post(
    "/training/start",
    response_model=TrainingJobPayload,
    status_code=status.HTTP_202_ACCEPTED,
)
async def start_training(payload: TrainingStartRequest, request: Request) -> dict[str, Any]:
    runtime = resolve_training_runtime(request)
    try:
        return await runtime.service.start_training(
            TrainingStartPayload(
                character_id=payload.character_id,
                mode=payload.mode,
                base_model=payload.base_model,
                confirm_flux_dev_license=payload.confirm_flux_dev_license,
                retrain_of_task_id=payload.retrain_of_task_id,
                retrain_step_delta=payload.retrain_step_delta,
            )
        )
    except TrainingCharacterNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TrainingValidationError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except TrainingServiceError as exc:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(exc)) from exc


@router.get("/training/tasks", response_model=list[TrainingJobPayload])
def list_training_tasks(
    request: Request,
    character_id: str | None = Query(default=None, alias="characterId"),
) -> list[dict[str, Any]]:
    runtime = resolve_training_runtime(request)
    return runtime.service.list_training_jobs(character_id=character_id)


@router.get("/training/tasks/{task_id}", response_model=TrainingJobPayload)
def get_training_task(task_id: str, request: Request) -> dict[str, Any]:
    runtime = resolve_training_runtime(request)
    try:
        return runtime.service.get_training_job(task_id)
    except TrainingJobNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
